### ISSUE TITLE
Jax2tf fails when mixing polymorphic types and constant indexing

### ISSUE THREAD

**Issue by gabisurita** [google/jax#6832]
The following snippet fails in the `tf_fn` call. There seems to be a check clause that is trying to multiply the integer index with the polymorphic DimVar.

```python
import jax
import numpy as np
from jax.experimental import jax2tf

def jax_fn(x):
    return x[:, 1] 

tf_fn = jax2tf.convert(jax_fn, polymorphic_shapes=["(n, 2)"])

tf_fn(np.ones((10, 2)))
```

The full traceback:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-d7ef2bb20a2d> in <module>
      8 tf_fn = jax2tf.convert(jax_fn, polymorphic_shapes=["(n, 2)"])
      9 
---> 10 tf_fn(np.ones((10, 2)))

~/.local/lib/python3.8/site-packages/jax/experimental/jax2tf/jax2tf.py in converted_fun(*args)
    284                   functools.partial(converted_grad_fn, _out_cts_avals=tuple(out_avals)))
    285 
--> 286         out_flat = converted_fun_flat_with_custom_gradient(*args_flat)
    287       else:
    288         out_flat_raw = _interpret_fun(flat_fun, args_flat, args_avals_flat)

~/.local/lib/python3.8/site-packages/tensorflow/python/ops/custom_gradient.py in __call__(self, *a, **k)
    259 
    260   def __call__(self, *a, **k):
--> 261     return self._d(self._f, a, k)
    262 
    263 

~/.local/lib/python3.8/site-packages/tensorflow/python/ops/custom_gradient.py in decorated(wrapped, args, kwargs)
    213 
    214     if context.executing_eagerly():
--> 215       return _eager_mode_decorator(wrapped, args, kwargs)
    216     else:
    217       return _graph_mode_decorator(wrapped, args, kwargs)

~/.local/lib/python3.8/site-packages/tensorflow/python/ops/custom_gradient.py in _eager_mode_decorator(f, args, kwargs)
    436   """Implement custom gradient decorator for eager mode."""
    437   with tape_lib.VariableWatcher() as variable_watcher:
--> 438     result, grad_fn = f(*args, **kwargs)
    439   args = nest.flatten(args)
    440   all_inputs = list(args) + list(kwargs.values())

~/.local/lib/python3.8/site-packages/jax/experimental/jax2tf/jax2tf.py in converted_fun_flat_with_custom_gradient(*args_flat)
    279         @tf.custom_gradient
    280         def converted_fun_flat_with_custom_gradient(*args_flat: TfVal) -> TfVal:
--> 281           out_with_avals = _interpret_fun(flat_fun, args_flat, args_avals_flat)
    282           outs, out_avals = util.unzip2(out_with_avals)
    283           return (tuple(outs),

~/.local/lib/python3.8/site-packages/jax/experimental/jax2tf/jax2tf.py in _interpret_fun(fun, in_vals, in_avals)
    311   with core.new_base_main(TensorFlowTrace) as main:  # type: ignore
    312     fun = _interpret_subtrace(fun, main, in_avals)
--> 313     out_vals: Sequence[Tuple[TfVal, core.AbstractValue]] = fun.call_wrapped(*in_vals)
    314     del main
    315   return tuple(out_vals)

~/.local/lib/python3.8/site-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    164 
    165     try:
--> 166       ans = self.f(*args, **dict(self.params, **kwargs))
    167     except:
    168       # Some transformations yield from inside context managers, so we have to

<ipython-input-1-d7ef2bb20a2d> in jax_fn(x)
      4 
      5 def jax_fn(x):
----> 6     return x[:, 1]
      7 
      8 tf_fn = jax2tf.convert(jax_fn, polymorphic_shapes=["(n, 2)"])

~/.local/lib/python3.8/site-packages/jax/core.py in __getitem__(self, idx)
    534   def __rshift__(self, other): return self.aval._rshift(self, other)
    535   def __rrshift__(self, other): return self.aval._rrshift(self, other)
--> 536   def __getitem__(self, idx): return self.aval._getitem(self, idx)
    537   def __nonzero__(self): return self.aval._nonzero(self)
    538   def __bool__(self): return self.aval._bool(self)

~/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py in _rewriting_take(arr, idx)
   4816   arr = asarray(arr)
   4817   treedef, static_idx, dynamic_idx = _split_index_for_jit(idx)
-> 4818   return _gather(arr, treedef, static_idx, dynamic_idx)
   4819 
   4820 # TODO(phawkins): re-enable jit after fixing excessive recompilation for

~/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py in _gather(arr, treedef, static_idx, dynamic_idx)
   4828   # Avoid calling gather if the slice shape is empty, both as a fast path and to
   4829   # handle cases like zeros(0)[array([], int32)].
-> 4830   if _prod(indexer.slice_shape) == 0:
   4831     return zeros_like(y, shape=indexer.slice_shape)
   4832 

~/.local/lib/python3.8/site-packages/jax/_src/util.py in prod(xs)
    207   out = 1
    208   for x in xs:
--> 209     out *= x
    210   return out
    211 

TypeError: unsupported operand type(s) for *=: 'int' and 'DimVar'
```

**Comment by gabisurita** [comment]
There seems to be a workaround using `lax.index_in_dim` directly. The problem only happens when using numpy style indexing.

```python
def jax_fn(x):
    return jax.lax.index_in_dim(x, 1, 1) 
```

**Comment by gnecula** [comment]
This particular use case of indexing will be fixed when we merge #6827.

However, it is good to keep in mind that the shape polymorphism feature of jax2tf is going to be incomplete in general: there are going to be programs that are shape polymorphic but jax2tf cannot prove it and will thus abort.

**Comment by gnecula** [comment]
I have verified that this works after merging #6827.
